src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java

```diff
@@ -115,27 +123,46 @@ public Object execute(
     if (METHOD_SET_READ_ONLY.equals(methodName) && args != null && args.length > 0) {
       switchConnectionIfRequired((Boolean) args[0]);
     }
+
     try {
-      return this.nextPlugin.execute(methodInvokeOn, methodName, executeSqlFunc, args);
+      final Object result = this.nextPlugin.execute(methodInvokeOn, methodName, executeSqlFunc, args);
+
+      if (this.explicitlyReadOnly && this.loadBalanceReadOnlyTraffic && isTransactionBoundary(methodName)) {
+        this.inTransaction = false;
+        pickNewReaderConnection();
+      }
+      return result;
     } catch (SQLException e) {
       if (isCommunicationException(e)) {
```

Comment:
Are we catching communication exceptions (any kind of them) or just failover exceptions? What happen if failover plugin isn't loaded?

Reply:
We're handling exceptions with SQLState of 08S02 (communication link changed, I believe this SQLState was created by us and is only thrown upon failover) and 08007 (transaction resolution unknown, which is thrown when we failover inside a transaction but probably is also thrown in other scenarios). I wanted to handle only failover exceptions but this seems like the closest we can get since I'm not sure of any other way to distinguish failover exceptions from other SQLExceptions. We don't have integration tests covering the scenario where the failover plugin isn't loaded, should we add tests for that? However, according to the code, if the failover plugin isn't loaded there are two scenarios:

1. The sql state is not 08007 (most cases), in which case the exception will just be thrown without doing anything
2. The sql state is 08007, in which case this block will be entered: all connections will be closed, the driver will attempt to update its topology, and updateInternalConnectionInfo will be called (which updates the current host index and potentially the writer/reader connection to point to the current connection if it is a writer/reader).